Bugfix/disable authentication

rickardbergatangrycreative opened the pull request:
Disabling authentication by adding an empty array to the filter "satispress_authentication_servers" currently doesn't work.
I've fixed that by adding the satispress_view_package- and satispress_download_package-capabilities to all users, if the filter "satispress_authentication_servers" returns an empty array.

thefrosty commented on src/ServiceProvider.php:
```diff
@@ -77,6 +77,15 @@ public function register( PimpleContainer $container ) {
 
+		$container['authentication.disable'] = function( $container ) {
+			$server_count = \iterator_count( $container['authentication.servers'] );
+			if ( 0 >= $server_count ) {
```
I would suggest just returning the count value rather than an if else statement. Example: `return 0 >= $server_count;

bradyvercher commented:
Thanks for the PR @rickardbergatangrycreative! I think the capabilities were added after settling on returning an empty array for the authentication servers to disable authentication, so I hadn't considered this.

Adding a new entry in the container seems like it could cause issues if the `authentication.disable` value were changed separately from the servers list. It also feels a little weird to be disabling authentication in the capabilities provider.

How about dropping the `authentication.disable` entry and moving the `user_has_cap` filter out of the `Capabilities` provider and into the  `Authentication` provider?

rickardbergatangrycreative commented:
@bradyvercher I agree, and I actually tried to come up with a solution with the `Authentication` provider, but couldn't figure it out for some reason and then got stuck with the idea of capabilities.

I've implemented your suggestion now.
I'm new to WordPress development, but hopefully I moved everything correctly.

bradyvercher commented:
Thanks for making that change and welcome to WordPress development! Feel free to reach out if you have any questions about anything.

There are a few additional minor things I'd like to see before merging this, so I'll leave a couple comments on the code.

bradyvercher commented on src/Provider/Authentication.php:
```diff
@@ -119,4 +121,15 @@ protected function get_request_path(): string {
+
+		if ( 0 >= $server_count ) {
+			$allcaps[Caps::DOWNLOAD_PACKAGES] = true;
+			$allcaps[Caps::VIEW_PACKAGES] = true;
```
Oh, and PHPCS will complain if the equal signs aren't aligned.

```php
$allcaps [ Caps::DOWNLOAD_PACKAGES ] = true;
$allcaps[ Caps::VIEW_PACKAGES ]      = true;
```

You can run `composer phpcs` to see any errors or warnings that it generates.

rickardbergatangrycreative replied:
I've added whitespace around the array indexes and aligned the equal signs.
I also ran `composer phpcs` afterwards which didn't find any errors after the changes.

rickardbergatangrycreative commented on src/Provider/Capabilities.php:
```diff
@@ -20,6 +20,7 @@
  * @since 0.3.0
  */
 class Capabilities extends AbstractHookProvider {
+
```
I removed the empty line.

rickardbergatangrycreative commented on src/Provider/Authentication.php:
```diff
@@ -119,4 +121,15 @@ protected function get_request_path(): string {
+	public function user_has_cap( array $allcaps, array $caps, array $args ) {
+		$server_count = \iterator_count( $this->servers );
+
+		if ( 0 >= $server_count ) {
```
I removed the `$server_count` variable and moved `iterator_count()` call into the conditional statement.

rickardbergatangrycreative commented on src/Provider/Authentication.php:
```diff
@@ -119,4 +121,15 @@ protected function get_request_path(): string {
 		return '/' . ltrim( $request_path, '/' );
 	}
+
+	public function user_has_cap( array $allcaps, array $caps, array $args ) {
```
The method has now been renamed to `maybe_allow_public_access`.
I also dropped the `$caps` and `$args` parameters as well as adjusted the number of parameters in the `add_filter()` call.

bradyvercher replied:
You can leave off the optional parameters in the `add_filter()` call since `10` and `1` are the defaults. Once that's updated, I'll go ahead and get this merged. Everything else looks good. Thanks!

rickardbergatangrycreative commented on src/Provider/Authentication.php:
```diff
@@ -119,4 +121,15 @@ protected function get_request_path(): string {
 
 		return '/' . ltrim( $request_path, '/' );
 	}
+
```
I wasn't entirely sure how much and what I needed to include, but I wrote a docblock stating what the method does as well as documenting what parameter it takes and what it returns.
